ignite/distributed/auto.py

```diff
@@ -16,7 +17,42 @@
 __all__ = ["auto_dataloader", "auto_model", "auto_optim", "DistributedProxySampler"]
 
 
-def auto_dataloader(dataset, **kwargs):
+if idist.has_xla_support:
+
+    import torch_xla.core.xla_model as xm
+    from torch_xla.distributed.parallel_loader import ParallelLoader
+
+    class _MpDeviceLoader:
+        # https://github.com/pytorch/xla/pull/2117
+        # From pytorch/xla if `torch_xla.distributed.parallel_loader.MpDeviceLoader` is not available
+        def __init__(self, loader: Any, device: torch.device, **kwargs: Any) -> None:
+            self._loader = loader
+            self._device = device
+            self._parallel_loader_kwargs = kwargs
+
+        def __setattr__(self, name: str, value: Any) -> None:
+            super().__setattr__(name, value)
+
+        def __getattr__(self, name: str) -> Any:
+            super().__getattribute__(name)
+
+        def __iter__(self) -> Iterator:
+            parallel_loader = ParallelLoader(self._loader, [self._device], **self._parallel_loader_kwargs)
+            return parallel_loader.per_device_loader(self._device)
+
+        def __len__(self) -> int:
+            return len(self._loader)
+
+    class _XLADistributedOptimizer(Optimizer):
+        def __init__(self, optimizer: Optimizer) -> None:
+            super(self.__class__, self).__init__(optimizer.param_groups, {})
+            self.wrapped_optimizer = optimizer
+
+        def step(self, closure: Optional[Callable] = None) -> None:
+            xm.optimizer_step(self.wrapped_optimizer, barrier=True)
+
+
+def auto_dataloader(dataset: Dataset, **kwargs: Any) -> Union[DataLoader, _MpDeviceLoader]:
```

Comment:
Can we use `Union[DataLoader, "_MpDeviceLoader"]` instead of moving all those private helper classes to the top ?

Reply:
yeah, I already did it locally.